core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -818,6 +859,62 @@ private static AnnotatedValueResolver ofInjectableTypes0(String name, AnnotatedE
         };
     }
 
+    /**
+     * Returns an attribute resolver which retrieves a value specified by {@code attrKeys}
+     * from the {@link RequestContext}.
+     */
+    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
+    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
+        return (resolver, ctx) -> {
+            final StringBuilder errorMsgBuilder = new StringBuilder();
+            for (AttributeKey<?> attrKey : attrKeys) {
+                final Object value = ctx.context.attr(attrKey);
+                if (value != null) {
+                    final boolean isValidType;
+
+                    final Class<?> rawContainerType = resolver.rawContainerType();
+                    if (rawContainerType != null) {
+                        isValidType = rawContainerType.isInstance(value);
+                        if (!isValidType) {
+                            errorMsgBuilder.append("The value ");
+                            errorMsgBuilder.append(value);
+                            errorMsgBuilder.append(" cannot be cast to rawContainerType ");
+                            errorMsgBuilder.append(rawContainerType);
+                        }
+                    } else {
+                        final Class<?> elementType = resolver.elementType().isPrimitive() ?
+                                                     Primitives.wrap(resolver.elementType())
+                                                     : resolver.elementType();
```

Comment:
We don't need to create a wrapped by for every iteration. How about?
```java
Class<?> targetType = resolver.rawContainerType();
if (targetType == null) {
    elementType = resolver.elementType().isPrimitive() ?
                     Primitives.wrap(resolver.elementType())
                     : resolver.elementType();
}

for (...) {
   final boolean isValidType = rawContainerType.isInstance(value);
   ...
   if (isValidType) {
      return value;
   } else {
      if (errorMsgBuilder == null) {
         errorMsgBuilder.append("... initial message")
      }
   }
}
```
The error message looks verbose a bit. Should we simplify like 
```
No matching value for '${resolver.httpElementName}' in the attributes: [{key=attr-key, value=attr-value}, {...}]`
```

Reply:
As our validation is not just a simple type casting, I prefer to raise `IllegalStateException` instead.

Reply:
@ikhoon nim, thanks for your review!! 🙇‍♂️ 

I refer to this comments https://github.com/line/armeria/pull/5547#discussion_r1570273226 from @jrhee17 nim! 

1. Use `rawType` instead of `rawContainerType`. thus i removed all point related with `rawContainerType`. 
2. and then `Class<?> targetType ` is more simple than before. because `rawType` can express `primitive` and `Reference Type` and `ContainerType`! 
3. String manipulation is getting too complicated because we trace all of `AttrKey`. to make simple, thrown information is only related with last value, as you recommend! (https://github.com/line/armeria/pull/5547#discussion_r1568345912)


When you have free time, could you please take a look again? 🙇‍♂️

Reply:
Had a quick check and the new changes look better. Let me take a look closer.